### PR TITLE
SONARJAVA-1242  Rule S3052: Fields should not be initialized to default values

### PR DESCRIPTION
**Pull request by mpaladin**
https://jira.sonarsource.com/browse/SONARJAVA-1242

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/DefaultInitializedFieldCheck.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    ExpressionTree initializer = member.initializer();
+    if (initializer != null) {
+      initializer = ExpressionsHelper.skipParentheses(initializer);
+      if (isDefault(initializer)) {
+        addIssue(member, "Remove this initialization to \"" + ((LiteralTree) initializer).value() + "\", the compiler will do that for you.");
+      }
+    }
+  }
+
+  private static boolean isDefault(ExpressionTree expression) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/DefaultInitializedFieldCheck.java]: MAJOR The Cyclomatic Complexity of this method "isDefault" is 15 which is greater than 10 authorized. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
MAJOR Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1142)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/checks/DefaultInitializedFieldCheck.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        return "'\\u0000'".equals(((LiteralTree) expression).value());
+      case BOOLEAN_LITERAL:
+        return "false".equals(((LiteralTree) expression).value());
+      case INT_LITERAL:
+      case LONG_LITERAL:
+        Long value = LiteralUtils.longLiteralValue(expression);
+        return value != null && value == 0;
+      case FLOAT_LITERAL:
+      case DOUBLE_LITERAL:
+        return Double.doubleToLongBits(Double.valueOf(((LiteralTree) expression).value())) == 0;
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/DefaultInitializedFieldCheck.java]: detail but I guess you can use this line only for all numerical types ?

Comment (mpaladin) [inline java-checks/src/main/java/org/sonar/java/checks/DefaultInitializedFieldCheck.java]: @benzonico doesn't like `L` suffix and hexadecimal integers.

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by axel3rd** [comment]
Hi, 

This PR ([SONARJAVA-1242](https://jira.sonarsource.com/browse/SONARJAVA-1242)) is in [v3.6](https://jira.sonarsource.com/projects/SONARJAVA?selectedItem=com.atlassian.jira.jira-projects-plugin:release-page&status=unreleased&contains=3.6), currently planned for end of week.
It provides a FP on final field initialization, fixed by #405.
If it could be reviewed for this version 3.6, it would be nice.
I'm available this week for any update if required.

Best regards.

**Pull request merged**